tests/unit/hook_packages/terraform/hooks/prepare/test_resource_linking.py

```diff
@@ -350,3 +357,227 @@ def test_resolve_module_variable_invalid_module_reference(
+    @patch("samcli.hook_packages.terraform.hooks.prepare.resource_linking._build_module_variables_from_configuration")
+    @patch("samcli.hook_packages.terraform.hooks.prepare.resource_linking._build_module_resources_from_configuration")
+    @patch("samcli.hook_packages.terraform.hooks.prepare.resource_linking._build_module_outputs_from_configuration")
+    @patch("samcli.hook_packages.terraform.hooks.prepare.resource_linking._build_child_modules_from_configuration")
+    def test_build_module(
+        self,
+        patched_build_child_modules_from_configuration,
+        patched_build_module_outputs_from_configuration,
+        patched_build_module_resources_from_configuration,
+        patched_build_module_variables_from_configuration,
+        patched_build_module_full_address,
+    ):
+        mock_full_address = Mock()
+        patched_build_module_full_address.return_value = mock_full_address
+
+        mock_variables = Mock()
+        patched_build_module_variables_from_configuration.return_value = mock_variables
+
+        mock_resources = Mock()
+        patched_build_module_resources_from_configuration.return_value = mock_resources
+
+        mock_outputs = Mock()
+        patched_build_module_outputs_from_configuration.return_value = mock_outputs
+
+        mock_child_modules = Mock()
+        patched_build_child_modules_from_configuration.return_value = mock_child_modules
+
+        result = _build_module(Mock(), Mock(), Mock(), Mock())
+        expected_module = TFModule(
+            mock_full_address, None, mock_variables, mock_resources, mock_child_modules, mock_outputs
+        )
+
+        self.assertEqual(result, expected_module)
+
+    @parameterized.expand(
+        [
+            (None, None, None),
+            ("some_module", None, "module.some_module"),
+            ("some_module", "parent_module_address", "parent_module_address.module.some_module"),
+        ]
+    )
+    def test_build_module_full_address(self, module_name, parent_module_address, expected_full_address):
+        result = _build_module_full_address(module_name, parent_module_address)
+        self.assertEqual(result, expected_full_address)
+
+    def test_build_module_variables_from_configuration(self):
+        module_configuration = {
+            "variables": {
+                "var1": {"default": "var1_default_value"},
+                "var2": {"default": "var2_default_value"},
+                "var3": {},
+                "var4": {},
+            },
+        }
+
+        input_variables = {
+            "var2": ConstantValue("var2_input_value"),
+            "var4": ConstantValue("var4_input_value"),
+        }
+
+        result = _build_module_variables_from_configuration(module_configuration, input_variables)
+
+        self.assertEqual(result["var1"], ConstantValue("var1_default_value"))
+        self.assertEqual(result["var2"], input_variables["var2"])
+        self.assertEqual(result["var3"], ConstantValue(None))
```

Comment:
why `ConstantValue(None)` and not `ConstantValue({})`

Reply:
this happens if a variable has no default value, as we only copy those. The `var3` variable doesn't have a `default` key in its object. There's no overriding value in input variables either, so it just stays as None. I think this is better than assuming some type like object, but let me know if you have concerns.

Reply:
ok got it .. my bad.